rfd/0005-kubernetes-service.md

```diff
@@ -0,0 +1,514 @@
+
+```yaml
+apiVersion: v1
+kind: Config
+clusters:
+- cluster:
+    server: https://a.example.com
+  name: a-cluster
+- cluster:
+    server: https://b.example.com
+  name: b-cluster
+contexts:
+- context:
+    cluster: a-cluster
+    user: a-creds
+  name: a
+- context:
+    cluster: b-cluster
+    user: b-creds
+  name: b
+current-context: a
+users:
+- name: a-creds
+  user:
+    client-key-data: "..."
+    client-certificate-data: "..."
+- name: b-creds
+  user:
+    client-key-data: "..."
+    client-certificate-data: "..."
+```
+
+Here, we have two k8s contexts. A proxy will refer to them as “a” and “b” from
+the `contexts` section, not “a-cluster” and “b-cluster” from the `clusters`
+section. An item in the `clusters` section doesn't specify which credentials to
+use on its own and there can be multiple credentials bound to the same cluster.
+
+### K8s cluster metadata
+
+With this, we need to make the clients aware of all the available k8s clusters.
+To do this, registered k8s services announce the k8s clusters via heartbeats.
+The `ServerSpecV2` gets a new field `kube_clusters` with a list of cluster
+names.
+
+Auth server and proxies can be queried for a union of all the k8s clusters
+announced by all the proxies.
+
+`tsh` uses this query to show all available k8s clusters to the user (`tsh kube
+clusters`).
+
+### Routing
+
+Now that the client knows about the available k8s clusters, they should be able
+to switch between them. Since we don’t control the actual k8s client
+(`kubectl`), our only transport to indicate the target k8s cluster is the user
+TLS certificate.
+
+The TLS certificates we issue for k8s access will include the k8s cluster name
+as an extension to the `Subject`, similar to how `kube_users` and `kube_groups`
+are embedded.
+
+This is similar to how the `RouteToCluster` field works today for routing to
+trusted clusters.
+
+When a proxy receives a k8s request, it does the following:
+- Authenticate the request (verify cert)
+- Parse target k8s cluster name from the cert extension
+- Check it against its local `kubernetes_service`
+- If found, forward the request and return
+- If not found, fetch the list of all k8s clusters from auth server
+- If k8s cluster is found in that list, forward the request to an appropriate address
+  - The target address could be a `kubernetes_service`
+  - Or another proxy that has a tunnel to the `kubernetes_service`
+- If k8s cluster is not found in that mapping, return an error
+
+### Service reverse tunnels
+
+Next, we need to support k8s clusters behind firewalls.
+
+Reusing the reverse tunnel logic from trusted clusters and IoT mode for nodes,
+we allow k8s service within the same cluster to tunnel to proxies. From the
+user perspective, it’s identical to IoT node setup: use a public proxy address
+in the `auth_servers` field of `teleport.yaml`.
+
+When connecting over a reverse tunnel, `kubernetes_service` will not listen on
+any local port, unless its `listen_addr` is set.
```

Comment:
I think this is different to the way that `ssh_service` works currently - I don't believe `ssh_service`  _ever_ listens when connected over a reverse tunnel, meaning there's no way to make an inbound connection to port 3022. I feel like the behaviour of `kubernetes_service` when connected over a reverse tunnel should be the same, unless there's a really compelling use case to make it listen locally?

Reply:
So you mean that `listen_addr` should be disallowed when connecting over a tunnel?
I can see 2 small (but not deal-breaker) problems:
- `listen_addr` is validated at startup, before the process knows whether it's talking to an auth server or a proxy
- an admin may choose to set `listen_addr` on all instances, but only set `auth_server` to a proxy on _some_ of them; on tunneled instances, listener will remain active but unused; on regular instances it will be dialed by the proxies; no need to template `listen_addr` in your config mangement

Reply:
As I start writing relevant code, I understand what you mean better.
I'll disallow both connecting via proxy tunnel and using a local `listen_addr` because:
- it's simpler to implement (no need to merge connections from multiple listeners)
- it behaves more like the `ssh_service`